### PR TITLE
Fix #4130: Unassignable param destructuring crash

### LINKED ISSUE #4130: Cannot read property 'length' of undefined
### Coffeescript:
(script.coffee)
```Coffeescript
fun = ({
  @param : null
}) ->
  console.log "Oh hello!"
```

### Compile:
```shell
./node_modules/.bin/coffee --no-header --bare --compile script.coffee
```

### Throws:
```terminal
TypeError: Cannot read property 'length' of undefined
  at Param.exports.Param.Param.eachName (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:2154:30)
  at Param.exports.Param.Param.eachName (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:2160:16)
  at Code.exports.Code.Code.eachParamName (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:2074:28)
  at Code.exports.Code.Code.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:2028:12)
  at Code.exports.Base.Base.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Block.exports.Block.Block.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:387:28)
  at Block.exports.Block.Block.compileWithDeclarations (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:480:19)
  at Code.exports.Code.Code.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:2055:63)
  at Code.exports.Base.Base.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Call.exports.Call.Call.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:1025:55)
  at Call.exports.Base.Base.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Parens.exports.Parens.Parens.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:2808:24)
  at Parens.exports.Base.Base.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Assign.exports.Assign.Assign.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:1706:24)
  at Assign.exports.Base.Base.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Class.exports.Class.Class.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:1624:20)
  at Class.exports.Base.Base.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Block.exports.Block.Block.compileNode (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:387:28)
  at Block.exports.Block.Block.compileWithDeclarations (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:480:19)
  at Block.exports.Block.Block.compileRoot (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:452:24)
  at Block.exports.Block.Block.compileToFragments (/.../app/node_modules/coffee-script/lib/coffee-script/nodes.js:369:21)
  at Object.<anonymous> (/.../app/node_modules/coffee-script/lib/coffee-script/coffee-script.js:64:38)
  at Object.compile (/.../app/node_modules/coffee-script/lib/coffee-script/coffee-script.js:34:19)
  at compileScript (/.../app/node_modules/coffee-script/lib/coffee-script/command.js:232:33)
  at compilePath (/.../app/node_modules/coffee-script/lib/coffee-script/command.js:174:14)
  at Object.exports.run (/.../app/node_modules/coffee-script/lib/coffee-script/command.js:98:20)
  at Object.<anonymous> (/.../app/node_modules/coffee-script/bin/coffee:7:41)
  at Module._compile (module.js:435:26)
  at Object.Module._extensions..js (module.js:442:10)
  at Module.load (module.js:356:32)
  at Function.Module._load (module.js:311:12)
  at Function.Module.runMain (module.js:467:10)
  at startup (node.js:134:18)
  at node.js:961:3
```
##### The same happens when if you replace `null` with:
- Booleans
- Functions
- Conditional values: e.g.: `if window? then true else false`

### But when you change `:` into `=`, it works:
```Coffeescript
fun = ({
  @param = null
})->
  console.log "Oh hello!"
```

(No extra terminal output)
### Generated JS:
```Javascript
var fun;

fun = function(arg) {
  var ref;
  this.param = (ref = arg.param) != null ? ref : null;
  return console.log("Oh hello!");
};
```

### PR DESCRIPTION
**Pull request by lydell**

### CODE REVIEW

File: test/error_messages.coffee [hunk 1]
```diff
@@ -825,6 +825,36 @@ test "#3926: implicit object in parameter list", ->
                                        ^
   '''
 
+test "#4130: unassignable in destrutured param", ->
```
Comment (bjmiller) [inline test/error_messages.coffee]: * destructured

**Comment by michaelficarra** [comment]
These tests don't include the particular repro case from the referenced issue. Is that intentional?

**Comment by lydell** [comment]
@michaelficarra Now it does :)

**Comment by michaelficarra** [comment]
Thanks @lydell.

**Pull request merged**